packages/shape-to-query/model/constraint/index.ts

```diff
@@ -31,4 +36,10 @@ export const constraintComponents = $rdf.termMap<Term, ConstraintComponentStatic
   [sh.ClassConstraintComponent, ClassConstraintComponent],
   [sh.NodeKindConstraintComponent, NodeKindConstraintComponent],
   [sh.LanguageInConstraintComponent, LanguageInConstraintComponent],
+  [sh.DatatypeConstraintComponent, DatatypeConstraintComponent],
+  [sh.MinInclusiveConstraintComponent, MinInclusiveConstraintComponent],
+  [sh.MaxInclusiveConstraintComponent, MaxInclusiveConstraintComponent],
+  [sh.MinExclusiveConstraintComponent, MinExclusiveConstraintComponent],
+  [sh.MaxExclusiveConstraintComponent, MaxExclusiveConstraintComponent],
+
```

Comment:
I think I would go for a simpler approach without subclasses.

Instead, you could have just the one type `RangeConstraintComponent`

```suggestion
  [sh.MinInclusiveConstraintComponent, RangeConstraintComponent],
  [sh.MaxInclusiveConstraintComponent, RangeConstraintComponent],
  [sh.MinExclusiveConstraintComponent, RangeConstraintComponent],
  [sh.MaxExclusiveConstraintComponent, RangeConstraintComponent],
```

Reply:
That would require a small change in `model/constraint/factory.ts`. The components are a map only to allow replacing them individually but otherwise, the implementation could be deduplicated first:

```js
const components = new Set([...constraintComponents.values()])
  return [...components].flatMap((component) => {
    return [...component.fromShape(shapeModel, factory)]
  })
```

This way we wouldn't need these copypaste subclasses